current/runtime/storage.md

```diff
@@ -56,38 +244,129 @@ decl_storage! {
+metadata associated with an object in [IPFS](https://ipfs.io/); this means that only the IPFS
+location (a hash that is bounded in size) needs to be stored on-chain.
+
+Hashes are only one mechanism that can be used to control the size of runtime storage. An example of
+another mechanism is [bounds](#Create-Bounds).
+
+### Verify First, Write Last
+
+The state of a blockchain network's storage is immutable; data can be changed, but there will always
+be a record of these changes, and making them typically incurs costs. Because of this, it is
+important that data is only persisted to runtime storage when it is certain that all preconditions
+have been met. In general, code blocks that may result in adding data to storage should be
+structured as follows:
+
+```rust
+{
+  // all checks and throwing code go here
+
+  // all storage writes go here; no throwing code below this line
+
+  // all event emissions go here
+}
+```
+
+Do not use runtime storage to store intermediate or transient data within the context of an
+operation that is logically atomic or data that will not be needed if the operation is to fail. This
+does not mean that runtime storage should not be used to track the state of ongoing actions that
+require multiple atomic operations, as in the case of
+[the multi-signature capabilities from the Utility pallet](https://crates.parity.io/pallet_utility/enum.Call.html#variant.as_multi).
+In this case, runtime storage is used to track the signatories on a dispatchable call even though a
+given call may never receive enough signatures to actually be invoked. In this case, each signature
+is considered an atomic event in the ongoing multi-signature operation; the data needed to record a
+single signature is not stored until after all the preconditions associated with that signature have
+been met.
+
+### Create Bounds
+
+Creating bounds on the size of storage items is an extremely effective way to control the use of
+runtime storage and one that is used repeatedly throughout the Substrate codebase. In general, any
+storage item whose size is determined by user action should have a bound on it.
+[The multi-signature capabilities from the Utility pallet](https://crates.parity.io/pallet_utility/trait.Trait.html#associatedtype.MaxSignatories)
+that were described above are one such example. In this case, the list of signatories associated
+with a multi-signature operation is provided by the multi-signature participants. Because this
+signatory list is [necessary to come to consensus](#What-to-Store) on the state of the
+multi-signature operation, it must be stored in the runtime. However, in order to give runtime
+developers control over how much space in storage these lists may occupy, the Utility pallet
+requires users to configure a bound on this number that will be included as a
+[precondition](#Verify-First-Write-Last) before anything is written to storage.
 
 ## Child Storage Tries
 
 TODO
 
+## Storage Cache
+
+TODO
+
```

Comment:
This may be good in the Advanced doc but don't think it's needed here. TBH I'm not sure how much runtime developers even need to know about these two things.